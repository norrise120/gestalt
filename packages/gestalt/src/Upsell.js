// @flow strict
import React, { type Element, type Node } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Box from './Box.js';
import Heading from './Heading.js';
import Icon from './Icon.js';
import IconButton from './IconButton.js';
import Image from './Image.js';
import Button from './Button.js';
import Text from './Text.js';
import { type AbstractEventHandler } from './AbstractEventHandler.js';
import styles from './Upsell.css';

type LinkData = {|
  accessibilityLabel?: string,
  href: string,
  label: string,
  onClick?: AbstractEventHandler<
    | SyntheticMouseEvent<HTMLButtonElement>
    | SyntheticMouseEvent<HTMLAnchorElement>
    | SyntheticKeyboardEvent<HTMLAnchorElement>
    | SyntheticKeyboardEvent<HTMLButtonElement>
  >,
|};

type Props = {|
  dismissButton?: {|
    accessibilityLabel: string,
    onDismiss: () => void,
  |},
  imageData?: {|
    component: Element<typeof Image | typeof Icon>,
    width?: number,
  |},
  message: string,
  primaryLink?: LinkData,
  secondaryLink?: LinkData,
  title?: string,
|};

const UpsellLink = ({
  data,
  stacked,
  type,
}: {|
  data: LinkData,
  stacked?: boolean,
  type: string,
|}): Node => {
  const color = type === 'primary' ? 'red' : 'gray';
  const { accessibilityLabel, href, label, onClick } = data;

  return (
    <Box
      display="block"
      smDisplay="flex"
      alignItems="center"
      justifyContent="center"
      paddingX={1}
      marginTop={type === 'primary' && stacked ? 2 : undefined}
      smMarginTop="auto"
      smMarginBottom="auto"
    >
      <Button
        accessibilityLabel={accessibilityLabel}
        color={color}
        href={href}
        onClick={onClick}
        role="link"
        size="lg"
        text={label}
      />
    </Box>
  );
};

export default function Upsell({
  dismissButton,
  imageData,
  message,
  primaryLink,
  secondaryLink,
  title,
}: Props): Node {
  const isImage = imageData?.component && imageData.component.type === Image;

  return (
    <Box
      display="flex"
      direction="column"
      smDirection="row"
      padding={6}
      smPadding={8}
      position="relative"
      rounding={4}
      borderStyle="shadow"
    >
      <Box
        smDisplay="flex"
        wrap
        width="100%"
        smMarginTop={-3}
        smMarginBottom={-3}
      >
        <Box
          display="flex"
          direction="column"
          smDirection="row"
          justifyContent="center"
          alignItems="center"
          marginBottom={primaryLink || secondaryLink ? 4 : undefined}
          smMarginBottom={primaryLink || secondaryLink ? 0 : undefined}
          smPaddingY={3}
        >
          {imageData && (
            <Box
              marginBottom={4}
              smMarginBottom={0}
              width={
                isImage ? Math.min(imageData.width || 128, 128) : undefined
              }
              flex="none"
            >
              {imageData.component}
            </Box>
          )}
          <Box maxWidth={648}>
            <Box
              display="flex"
              smDisplay="block"
              direction="column"
              alignItems="center"
              marginBottom="auto"
              marginTop="auto"
              marginEnd={0}
              marginStart={0}
              smMarginEnd={6}
              smMarginStart={imageData ? 6 : 0}
            >
              {/* We repeat this code block to ensure that text is 
              centered for our smaller displays and left aligned 
              for larger displays */}
              <Box smDisplay="none">
                {title && (
                  <Box marginBottom={2}>
                    <Heading align="center" size="sm">
                      {title}
                    </Heading>
                  </Box>
                )}
                <Text align="center">{message}</Text>
              </Box>
              <Box smDisplay="block" display="none">
                {title && (
                  <Box marginBottom={2}>
                    <Heading size="sm">{title}</Heading>
                  </Box>
                )}
                <Text>{message}</Text>
              </Box>
            </Box>
          </Box>
        </Box>
        <Box smDisplay="flex" marginStart="auto" smMarginEnd={4} smPaddingY={3}>
          {secondaryLink && (
            <UpsellLink type="secondary" data={secondaryLink} />
          )}
          {primaryLink && (
            <UpsellLink
              stacked={!!secondaryLink}
              type="primary"
              data={primaryLink}
            />
          )}
        </Box>
      </Box>
      {dismissButton && (
        <div className={classnames(styles.rtlPos)}>
          <IconButton
            accessibilityLabel={dismissButton.accessibilityLabel}
            icon="cancel"
            iconColor="darkGray"
            onClick={dismissButton.onDismiss}
            padding={4}
            size="sm"
          />
        </div>
      )}
    </Box>
  );
}

Upsell.propTypes = {
  // $FlowFixMe[signature-verification-failure] flow 0.135.0 upgrade
  dismissButton: PropTypes.shape({
    accessibilityLabel: PropTypes.string.isRequired,
    onDismiss: PropTypes.func.isRequired,
  }),
  // $FlowFixMe[signature-verification-failure] flow 0.135.0 upgrade
  imageData: PropTypes.shape({
    component: PropTypes.node.isRequired,
    width: PropTypes.number,
  }),
  message: PropTypes.string.isRequired,
  // $FlowFixMe[signature-verification-failure] flow 0.135.0 upgrade
  primaryLink: PropTypes.shape({
    href: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    onClick: PropTypes.func,
    accessibilityLabel: PropTypes.string,
  }),
  // $FlowFixMe[signature-verification-failure] flow 0.135.0 upgrade
  secondaryLink: PropTypes.shape({
    href: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    onClick: PropTypes.func,
    accessibilityLabel: PropTypes.string,
  }),
  title: PropTypes.string,
};
