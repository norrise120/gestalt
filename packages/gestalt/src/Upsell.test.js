// @flow strict
import React from 'react';
import { create } from 'react-test-renderer';
import Icon from './Icon.js';
import Upsell from './Upsell.js';

describe('<Upsell />', () => {
  test('Basic Upsell', () => {
    const tree = create(
      <Upsell message="Insert a clever upsell message here" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('message + title', () => {
    const tree = create(
      <Upsell message="Insert a clever upsell message here" title="A Title" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('message + title + primaryLink', () => {
    const tree = create(
      <Upsell
        message="Insert a clever upsell message here"
        primaryLink={{ href: 'pinterest.com', label: 'Visit Pinterest' }}
        title="A Title"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('message + title + primaryLink + secondaryLink', () => {
    const tree = create(
      <Upsell
        message="Insert a clever upsell message here"
        primaryLink={{ href: 'pinterest.com', label: 'Visit Pinterest' }}
        secondaryLink={{ href: 'pinterest.com/help', label: 'Learn more' }}
        title="A Title"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('message + title + primaryLink + dismissButton', () => {
    const tree = create(
      <Upsell
        message="Insert a clever upsell message here"
        primaryLink={{ href: 'pinterest.com', label: 'Visit Pinterest' }}
        dismissButton={{
          accessibilityLabel: 'Dismiss banner',
          onDismiss: () => {},
        }}
        title="A Title"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('message + title + primaryLink + dismissButton + image', () => {
    const tree = create(
      <Upsell
        message="Insert a clever upsell message here"
        primaryLink={{ href: 'pinterest.com', label: 'Visit Pinterest' }}
        dismissButton={{
          accessibilityLabel: 'Dismiss banner',
          onDismiss: () => {},
        }}
        title="A Title"
        image={
          <Icon
            icon="pinterest"
            accessibilityLabel="Pin"
            color="darkGray"
            size={32}
          />
        }
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
