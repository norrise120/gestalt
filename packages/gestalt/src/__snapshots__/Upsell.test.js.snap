// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Upsell /> Basic Upsell 1`] = `
<div
  className="box mdDirectionRow mdPaddingX8 mdPaddingY8 paddingX6 paddingY6 relative rounding4 shadow xsDirectionColumn xsDisplayFlex"
>
  <div
    className="box flexWrap mdDisplayFlex mdMarginBottomN3 mdMarginTopN3"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="box mdPaddingY3 xsDisplayFlex"
    >
      <div
        className="box marginBottomAuto marginEnd6 marginTopAuto"
        style={
          Object {
            "maxWidth": 648,
          }
        }
      >
        <div
          className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
        >
          Insert a clever upsell message here
        </div>
      </div>
    </div>
    <div
      className="box marginStartAuto mdDisplayFlex mdMarginEnd4 mdPaddingY3"
    />
  </div>
</div>
`;

exports[`<Upsell /> message + title + primaryLink + dismissButton 1`] = `
<div
  className="box mdDirectionRow mdPaddingX8 mdPaddingY8 paddingX6 paddingY6 relative rounding4 shadow xsDirectionColumn xsDisplayFlex"
>
  <div
    className="box flexWrap mdDisplayFlex mdMarginBottomN3 mdMarginTopN3"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="box marginBottom4 mdMarginBottom0 mdPaddingY3 xsDisplayFlex"
    >
      <div
        className="box marginBottomAuto marginEnd6 marginTopAuto"
        style={
          Object {
            "maxWidth": 648,
          }
        }
      >
        <div
          className="box marginBottom2"
        >
          <h3
            className="Heading fontSize1 darkGray alignLeft breakWord"
          >
            A Title
          </h3>
        </div>
        <div
          className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
        >
          Insert a clever upsell message here
        </div>
      </div>
    </div>
    <div
      className="box marginStartAuto mdDisplayFlex mdMarginEnd4 mdPaddingY3"
    >
      <div
        className="box itemsCenter justifyCenter mdMarginBottomAuto mdMarginTopAuto paddingX1 xsDisplayFlex"
      >
        <a
          className="link hideOutline tapTransition block pill accessibilityOutline buttonLink button lg red"
          href="pinterest.com"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={null}
          onFocus={[Function]}
          onKeyPress={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          rel=""
          tabIndex={0}
          target={null}
        >
          <div
            className="Text fontSize3 white alignCenter fontWeightBold"
          >
            Visit Pinterest
          </div>
        </a>
      </div>
    </div>
  </div>
  <div
    className="rtlPos"
  >
    <button
      aria-label="Dismiss banner"
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 48,
            "width": 48,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={16}
          role="img"
          viewBox="0 0 24 24"
          width={16}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`<Upsell /> message + title + primaryLink + secondaryLink 1`] = `
<div
  className="box mdDirectionRow mdPaddingX8 mdPaddingY8 paddingX6 paddingY6 relative rounding4 shadow xsDirectionColumn xsDisplayFlex"
>
  <div
    className="box flexWrap mdDisplayFlex mdMarginBottomN3 mdMarginTopN3"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="box marginBottom4 mdMarginBottom0 mdPaddingY3 xsDisplayFlex"
    >
      <div
        className="box marginBottomAuto marginEnd6 marginTopAuto"
        style={
          Object {
            "maxWidth": 648,
          }
        }
      >
        <div
          className="box marginBottom2"
        >
          <h3
            className="Heading fontSize1 darkGray alignLeft breakWord"
          >
            A Title
          </h3>
        </div>
        <div
          className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
        >
          Insert a clever upsell message here
        </div>
      </div>
    </div>
    <div
      className="box marginStartAuto mdDisplayFlex mdMarginEnd4 mdPaddingY3"
    >
      <div
        className="box itemsCenter justifyCenter mdMarginBottomAuto mdMarginTopAuto paddingX1 xsDisplayFlex"
      >
        <a
          className="link hideOutline tapTransition block pill accessibilityOutline buttonLink button lg gray"
          href="pinterest.com/help"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={null}
          onFocus={[Function]}
          onKeyPress={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          rel=""
          tabIndex={0}
          target={null}
        >
          <div
            className="Text fontSize3 darkGray alignCenter fontWeightBold"
          >
            Learn more
          </div>
        </a>
      </div>
      <div
        className="box itemsCenter justifyCenter marginTop2 mdMarginBottomAuto mdMarginTopAuto paddingX1 xsDisplayFlex"
      >
        <a
          className="link hideOutline tapTransition block pill accessibilityOutline buttonLink button lg red"
          href="pinterest.com"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={null}
          onFocus={[Function]}
          onKeyPress={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          rel=""
          tabIndex={0}
          target={null}
        >
          <div
            className="Text fontSize3 white alignCenter fontWeightBold"
          >
            Visit Pinterest
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Upsell /> message + title + primaryLink 1`] = `
<div
  className="box mdDirectionRow mdPaddingX8 mdPaddingY8 paddingX6 paddingY6 relative rounding4 shadow xsDirectionColumn xsDisplayFlex"
>
  <div
    className="box flexWrap mdDisplayFlex mdMarginBottomN3 mdMarginTopN3"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="box marginBottom4 mdMarginBottom0 mdPaddingY3 xsDisplayFlex"
    >
      <div
        className="box marginBottomAuto marginEnd6 marginTopAuto"
        style={
          Object {
            "maxWidth": 648,
          }
        }
      >
        <div
          className="box marginBottom2"
        >
          <h3
            className="Heading fontSize1 darkGray alignLeft breakWord"
          >
            A Title
          </h3>
        </div>
        <div
          className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
        >
          Insert a clever upsell message here
        </div>
      </div>
    </div>
    <div
      className="box marginStartAuto mdDisplayFlex mdMarginEnd4 mdPaddingY3"
    >
      <div
        className="box itemsCenter justifyCenter mdMarginBottomAuto mdMarginTopAuto paddingX1 xsDisplayFlex"
      >
        <a
          className="link hideOutline tapTransition block pill accessibilityOutline buttonLink button lg red"
          href="pinterest.com"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={null}
          onFocus={[Function]}
          onKeyPress={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          rel=""
          tabIndex={0}
          target={null}
        >
          <div
            className="Text fontSize3 white alignCenter fontWeightBold"
          >
            Visit Pinterest
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Upsell /> message + title 1`] = `
<div
  className="box mdDirectionRow mdPaddingX8 mdPaddingY8 paddingX6 paddingY6 relative rounding4 shadow xsDirectionColumn xsDisplayFlex"
>
  <div
    className="box flexWrap mdDisplayFlex mdMarginBottomN3 mdMarginTopN3"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="box mdPaddingY3 xsDisplayFlex"
    >
      <div
        className="box marginBottomAuto marginEnd6 marginTopAuto"
        style={
          Object {
            "maxWidth": 648,
          }
        }
      >
        <div
          className="box marginBottom2"
        >
          <h3
            className="Heading fontSize1 darkGray alignLeft breakWord"
          >
            A Title
          </h3>
        </div>
        <div
          className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
        >
          Insert a clever upsell message here
        </div>
      </div>
    </div>
    <div
      className="box marginStartAuto mdDisplayFlex mdMarginEnd4 mdPaddingY3"
    />
  </div>
</div>
`;
